File: day19/python/tubes.py
#!/usr/bin/python3

import sys
from enum import Enum

class Direction(Enum):
    Up = 0,
    Down = 1,
    Right = 2,
    Left = 3


if len(sys.argv) != 2:
    print('Provide input file as command line argument!')
else:
    diagram = open(sys.argv[1]).read().splitlines()
    direction = Direction.Down
    posX, posY = diagram[0].index('|'), 0
    letters = []
    steps = 0

    while diagram[posY][posX] != ' ':
        steps += 1

        if diagram[posY][posX] == '+': # crossing
            if direction == Direction.Right or direction == Direction.Left:
                if diagram[posY + 1][posX] != ' ':
                    direction = Direction.Down
                    posY += 1
                elif diagram[posY - 1][posX] != ' ':
                    direction = Direction.Up
                    posY -= 1
            elif direction == Direction.Up or direction == Direction.Down:
                if diagram[posY][posX + 1] != ' ':
                    direction = Direction.Right
                    posX += 1
                elif diagram[posY][posX - 1] != ' ':
                    direction = Direction.Left
                    posX -= 1
        else:
            if diagram[posY][posX].isalpha(): # collect letters
                letters.append(diagram[posY][posX])

            if   direction == Direction.Down:  posY += 1
            elif direction == Direction.Up:    posY -= 1
            elif direction == Direction.Right: posX += 1
            elif direction == Direction.Left:  posX -= 1

    print('Collected letters:', ''.join(letters))
    print('Steps taken:', steps)

File: day19/rust/tubes.rs
use std::fs;

#[derive(Eq, PartialEq)]
enum Direction {
    Up,
    Down,
    Left,
    Right,
}

fn main() {
    let content = fs::read_to_string("../input").expect("file not found");
    let diagram: Vec<&str> = content.lines().collect();
    let mut position = (diagram[0].find('|').unwrap(), 0);
    let mut direction = Direction::Down;
    let mut letters = Vec::new();
    let mut steps = 0;

    while !diagram[position.1]
        .chars()
        .nth(position.0)
        .unwrap()
        .is_whitespace()
    {
        let character = diagram[position.1].chars().nth(position.0).unwrap();
        if character.is_alphabetic() {
            letters.push(character);
        } else if character == '+' {
            if direction == Direction::Right || direction == Direction::Left {
                if !diagram[position.1 + 1]
                    .chars()
                    .nth(position.0)
                    .unwrap()
                    .is_whitespace()
                {
                    direction = Direction::Down;
                } else if !diagram[position.1 - 1]
                    .chars()
                    .nth(position.0)
                    .unwrap()
                    .is_whitespace()
                {
                    direction = Direction::Up;
                }
            } else if direction == Direction::Up || direction == Direction::Down {
                if !diagram[position.1]
                    .chars()
                    .nth(position.0 + 1)
                    .unwrap()
                    .is_whitespace()
                {
                    direction = Direction::Right;
                } else if !diagram[position.1]
                    .chars()
                    .nth(position.0 - 1)
                    .unwrap()
                    .is_whitespace()
                {
                    direction = Direction::Left;
                }
            }
        }

        position = match direction {
            Direction::Up => (position.0, position.1 - 1),
            Direction::Down => (position.0, position.1 + 1),
            Direction::Left => (position.0 - 1, position.1),
            Direction::Right => (position.0 + 1, position.1),
        };
        steps += 1;
    }

    println!("{}", letters.iter().collect::<String>());
    println!("{}", steps);
}
